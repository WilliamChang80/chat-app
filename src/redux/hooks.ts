import { useSelector } from "react-redux";

export const useAppSelector = useSelector;

export function withPayloadType<T>() {
  return (t: T) => ({ payload: t });
}
