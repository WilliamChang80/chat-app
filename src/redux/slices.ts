import {
  createSlice,
  PayloadAction,
  SliceCaseReducers,
  ValidateSliceCaseReducers,
} from "@reduxjs/toolkit";

interface GenericState<T> {
  data?: T;
  status: "loading" | "finished" | "error";
}

interface GenericStateProps<T, J extends SliceCaseReducers<GenericState<T>>> {
  name: string;
  initialState: GenericState<T>;
  reducers: ValidateSliceCaseReducers<GenericState<T>, J>;
}

export const createGenericSlice = <
  T,
  Reducers extends SliceCaseReducers<GenericState<T>>
>({
  name = "",
  initialState,
  reducers,
}: GenericStateProps<T, Reducers>) =>
  createSlice({
    name,
    initialState,
    reducers: {
      start(state) {
        state.status = "loading";
      },
      success(state: GenericState<T>, action: PayloadAction<T>) {
        state.data = action.payload;
        state.status = "finished";
      },
      ...reducers,
    },
  });
