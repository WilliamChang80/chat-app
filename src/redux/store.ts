import { configureStore } from "@reduxjs/toolkit";
import counterReducer from "redux/features/counter/slices";

const store = configureStore({
  reducer: {
    counter: counterReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
