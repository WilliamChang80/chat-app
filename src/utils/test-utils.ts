import { ReactTestInstance, ReactTestRenderer } from "react-test-renderer";

export const findByTestId = (
  instance: ReactTestRenderer,
  testId: string
): string | ReactTestInstance =>
  instance.root.findByProps({ "data-testid": testId }).children[0];
