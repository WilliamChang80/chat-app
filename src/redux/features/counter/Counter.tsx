import { RootState } from 'redux/store'
import { connect } from 'react-redux'
import * as counterActions from './slices'
import { AnyAction, Dispatch } from 'redux'

interface CounterDispatchProps {
  increment: VoidFunction,
  decrement: VoidFunction,
  incrementByNumber: (count: number) => void,
}

interface CounterStateProps {
  counter: counterActions.CounterState
}

type CounterProps = CounterDispatchProps&CounterStateProps

const Counter = ({increment, decrement, incrementByNumber, counter}: CounterProps) => {
  return (
    <div>
      <div>
        <button
          aria-label="Decrement value"
          onClick={() => decrement()}
        >
          Decrement
        </button>
        <span>{counter.count}</span>
        <button
          aria-label="Increment value"
          onClick={() => increment()}
        >
          Increment
        </button>
        <button
          aria-label="+10"
          onClick={() => incrementByNumber(10)}
        >
          +10
        </button>
      </div>
    </div>
  )
}

const mapStateToProps = ({counter}: RootState)  => {
  return {
    counter
  }
}

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>): CounterDispatchProps => {
  return {
    increment: () => dispatch(counterActions.increment()),
    decrement: () => dispatch(counterActions.decrement()),
    incrementByNumber: (count: number) => dispatch(counterActions.incrementByNumber(count))
  } 
}

export default connect(mapStateToProps, mapDispatchToProps)(Counter);