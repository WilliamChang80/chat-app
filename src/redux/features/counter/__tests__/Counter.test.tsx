import { Provider } from "react-redux";
import { create, ReactTestRenderer } from "react-test-renderer";
import configureStore from "redux-mock-store";
import { RootState } from "redux/store";
import { findByTestId } from "../../../../utils/test-utils";
import Counter from "../Counter";

describe("test counter", () => {
  const mockStore = configureStore([]);

  it("increase when + button clicked", () => {
    const rootState: RootState = {
      counter: { count: 1 },
    };

    const store = mockStore(rootState);

    const component: ReactTestRenderer = create(
      <Provider store={store}>
        <Counter />
      </Provider>
    );

    expect(findByTestId(component, "counter-value")).toEqual("1");
  });
});
