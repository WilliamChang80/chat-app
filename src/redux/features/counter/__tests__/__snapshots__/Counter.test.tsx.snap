// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`increase when + button clicked 1`] = `
<div>
  <div>
    <button
      aria-label="Decrement value"
      onClick={[Function]}
    >
      Decrement
    </button>
    <span>
      1
    </span>
    <button
      aria-label="Increment value"
      onClick={[Function]}
    >
      Increment
    </button>
    <button
      aria-label="+10"
      onClick={[Function]}
    >
      +10
    </button>
  </div>
</div>
`;

exports[`test counter increase when + button clicked 1`] = `
<div>
  <div>
    <button
      aria-label="Decrement value"
      onClick={[Function]}
    >
      Decrement
    </button>
    <span
      data-testid="counter-value"
    >
      1
    </span>
    <button
      aria-label="Increment value"
      onClick={[Function]}
    >
      Increment
    </button>
    <button
      aria-label="+10"
      onClick={[Function]}
    >
      +10
    </button>
  </div>
</div>
`;
